### PR TITLE
Replace echo package with Go http.ServeMux

### PR DESCRIPTION
**Pull request by daniel-nichter**
This PR rewrites `api/api.go`. To avoid a large messy diff, I rewrote that code as `api/api_gomux.go` and deleted `api/api.go`. Functionally, the new code is identical. All tests pass.

Changes:
* Replace echo pkg with Go http.ServeMux
* Update packages (go.mod)
* Add test/docker/docker-compose.yml (for local testing)
* Add `api.ErrEndpointNotFound` to replace Echo's built-in 404 "Not found" error for nonexistent endpoints
* Add empty entity check to validator (instead of checking separately in handler)
* Add make-docs.sh (Go Swagger usage isn't obvious)

### CODE REVIEW

File: test/db.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 )
 
 var (
-	url      = "localhost:27017"
+	url      = "localhost:27017/?replicaSet=rs0&directConnection=true"
```
Comment (jemiahw) [inline test/db.go:15]: Please update the readme with instructions on test setup (how to run the docker container and any other prereqs)

**Review by jemiahw — COMMENTED** [review]
lgtm. I put a few comments inline but I don't see any show stoppers. 

api_gomux.go is a big file. Hard to review line by line for correctness, but I did go through it and overall it looks good. I ran tests and code coverage is good.

currently the tests run with the test docker image, but the server will not (by default). Should we update so that both the tests and server default to the same configuration which then works with the test docker image? Or is intentional to keep them separate so that test runs don't pollute the default server?

**Comment by daniel-nichter** [comment]
> currently the tests run with the test docker image, but the server will not (by default). Should we update so that both the tests and server default to the same configuration which then works with the test docker image? Or is intentional to keep them separate so that test runs don't pollute the default server?

Server won't? It defaults to `"mongodb://localhost:27017"`, so it should work. I'll try it...

File: docs/docs.go [hunk 2]
```diff
@@ -74,182 +51,6 @@ var doc = `{
...
-                    },
-                    "404": {
-                        "description": "Not Found",
-                        "schema": {
-                            "$ref": "#/definitions/etre.Error"
-                        }
-                    }
-                }
-            },
-            "put": {
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply docs/docs.go:95]: Somewhat intentional: the docs before were (ironically) undocumented and relied on an Echo-specific pkg. Making them work again wasn't easy--Swagger is surprisingly difficult to set up, imho, at least with Go. So for now I got the basics working again and I was intended to add back the docs (once I re-learn enough Swagger syntax) in another PR.

File: api/single_entity_write_test.go [hunk 3]
```diff
@@ -681,7 +681,7 @@ func TestPutEntityErrors(t *testing.T) {
 
 	if gotWR.Error == nil {
 		t.Errorf("WriteResult.Error is nil, expected error message")
-	} else if gotWR.Error.Type != "no-content" {
+	} else if gotWR.Error.Type != "empty-entity" {
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply api/single_entity_write_test.go:684]: New one is more accurate: the test sends `{}` which means HTTP content len = 2, but it's an empty entity. So rather than change the test input, I added a new error. The no-content error is still possible if caller does a request with a true zero length HTTP body.

File: test/db.go (hunk 1 above)
Comment (daniel-nichter) [reply test/db.go:15]: Done in ad0f809783b0275dfc5f062ba4c81d4bfe69609a.

File: api/single_entity_write_test.go [hunk 4]
```diff
@@ -681,7 +681,7 @@ func TestPutEntityErrors(t *testing.T) {
...
 		t.Errorf("WriteResult.Error.Type = %s, expected no-content", gotWR.Error.Type)
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply api/single_entity_write_test.go]: Done in ad0f809783b0275dfc5f062ba4c81d4bfe69609a.

File: api/api_gomux.go [hunk 5]
```diff
@@ -0,0 +1,1291 @@
...
+				}
+				return
+			}
+			ctx, cancel = context.WithTimeout(context.Background(), d)
+		}
+		defer cancel() // don't leak
+		t0 := time.Now()
+
+		// --------------------------------------------------------------
+		// Authenticate
```
[the remark this answers is not in the document]
Comment (daniel-nichter) [reply api/api_gomux.go:251]: I often thought the same while (re)writing but kept it inline for two reasons: 1) practicing "deep abstractions" and 2) the only endpoints that don't auth and the rest are metrics and docs, so 99% of routes would need an extra `auth()` middleware wrapper, only 2 or so wouldn't.

**Review by jemiahw — APPROVED** [review]
